src/brod_transaction.erl

```diff
@@ -0,0 +1,412 @@
+-type transactional_id() :: kpro:transactional_id().
+-type txn_ctx() :: kpro:txn_ctx().
+-type value() :: brod:value().
+-type batch_input() :: kpro:batch_input().
+
+
+-record(state,
+        { client_pid      :: client()
+        , context         :: txn_ctx()
+        , timeout         :: pos_integer()
+        , sequences       :: map()
+        , sent_partitions :: map()
+        , max_retries     :: pos_integer()
+        , backoff_step    :: pos_integer()
+        }).
+
+-type state() :: #state{}.
+
+init({Client, TxId, PropListConfig}) ->
+  ClientPid = pid(Client),
+  erlang:process_flag(trap_exit, true),
+
+  Config =
+  #{ max_retries := MaxRetries
+   , backoff_step := BackOffStep
+   , timeout := Timeout} = maps:merge(#{ max_retries => 5
+                                       , backoff_step => 100
+                                       , timeout => 1000},
+                                      proplists:to_map(PropListConfig)),
+
+  {ok, CTX} = make_txn_context(ClientPid, TxId, Config),
+  {ok, #state{ client_pid = ClientPid
+             , context = CTX
+             , max_retries = MaxRetries
+             , backoff_step= BackOffStep
+             , timeout = Timeout
+             , sequences = #{}
+             , sent_partitions = #{}
+             }}.
+
+%%% @private
+handle_call({add_offsets, ConsumerGroup, Offsets}, _From,
+            #state{ client_pid = Client
+                  , context = CTX
+                  , max_retries = MaxRetries
+                  , backoff_step = BackOffStep
+                  } = State) ->
+  Resp = do_add_offsets(Client, CTX, ConsumerGroup, Offsets,
+                        #{ max_retries => MaxRetries
+                         , backoff_step => BackOffStep}),
+  {reply, Resp, State};
+
+%%% @private
+handle_call({produce, Topic, Partition, Key, Value}, _From,
+            #state{} = OldState) ->
+  case do_produce(Topic, Partition, Key, Value, OldState) of
+    {ok, {Offset, State}} -> {reply, {ok, Offset}, State};
+    {error, Reason} -> {reply, {error, Reason}, OldState}
+  end;
+
+%%% @private
+handle_call({produce, Topic, Partition, Batch}, _From,
+            #state{} = OldState) ->
+  case do_batch_produce(Topic, Partition, Batch, OldState) of
+    {ok, {Offset, State}} -> {reply, {ok, Offset}, State};
+    {error, Reason} -> {reply, {error, Reason}, OldState}
+  end;
+
+%% @private
+handle_call(commit, _From, #state{context = CTX} = State) ->
+  {stop, normal, kpro:txn_commit(CTX), State};
+
+%% @private
+handle_call(terminate, _From, State) ->
+  {stop, normal, ok, #state{} = State};
+
+%% @private
+handle_call(abort, _From,
+            #state{context = CTX} = State) ->
+  {stop, normal, kpro:txn_abort(CTX), State};
+
+%% @private
+handle_call({abort, Timeout}, _From,
+            #state{context = CTX} = State) ->
+  {stop, normal,
+   kpro:txn_abort(CTX, #{timeout => Timeout}),
+   State};
+
+%% @private
+handle_call(stop, _From, #state{} = State) ->
+  {stop, normal, ok, State};
```

Comment:
Coding style: we usually don't emphasize individual clauses of a function and group them together (unlike Elixir):
```suggestion
%%% @private
handle_call({add_offsets, ConsumerGroup, Offsets}, _From,
            #state{ client_pid = Client
                  , context = CTX
                  , max_retries = MaxRetries
                  , backoff_step = BackOffStep
                  } = State) ->
  Resp = do_add_offsets(Client, CTX, ConsumerGroup, Offsets,
                        #{ max_retries => MaxRetries
                         , backoff_step => BackOffStep}),
  {reply, Resp, State};
handle_call({produce, Topic, Partition, Key, Value}, _From,
            #state{} = OldState) ->
  case do_produce(Topic, Partition, Key, Value, OldState) of
    {ok, {Offset, State}} -> {reply, {ok, Offset}, State};
    {error, Reason} -> {reply, {error, Reason}, OldState}
  end;
handle_call({produce, Topic, Partition, Batch}, _From,
            #state{} = OldState) ->
  case do_batch_produce(Topic, Partition, Batch, OldState) of
    {ok, {Offset, State}} -> {reply, {ok, Offset}, State};
    {error, Reason} -> {reply, {error, Reason}, OldState}
  end;
handle_call(commit, _From, #state{context = CTX} = State) ->
  {stop, normal, kpro:txn_commit(CTX), State};
handle_call(terminate, _From, State) ->
  {stop, normal, ok, #state{} = State};
handle_call(abort, _From,
            #state{context = CTX} = State) ->
  {stop, normal, kpro:txn_abort(CTX), State};
handle_call({abort, Timeout}, _From,
            #state{context = CTX} = State) ->
  {stop, normal,
   kpro:txn_abort(CTX, #{timeout => Timeout}),
   State};
handle_call(stop, _From, #state{} = State) ->
  {stop, normal, ok, State};
```